lib/extractor/layer.ts

```diff
@@ -53,6 +56,73 @@ export async function extractDockerArchive(
   });
 }
 
+export async function extractOciArchive(
+  ociArchiveFilesystemPath: string,
+  extractActions: ExtractAction[],
+): Promise<{
+  layers: ExtractedLayers[];
+  manifest: OciArchiveManifest;
+}> {
+  return new Promise((resolve, reject) => {
+    const tarExtractor: Extract = extract();
+
+    const layers: Record<string, ExtractedLayers> = {};
+    const manifests: Record<string, OciArchiveManifest> = {};
+    let imageIndex: OciImageIndex | undefined;
+
+    tarExtractor.on("entry", async (header, stream, next) => {
+      if (header.type === "file") {
+        if (isImageIndexFile(header.name)) {
+          imageIndex = await streamToJson<OciImageIndex>(stream);
+        } else {
+          const jsonStream = new PassThrough();
+          const layerStream = new PassThrough();
+          stream.pipe(jsonStream);
+          stream.pipe(layerStream);
+
+          const promises = [
+            streamToJson(jsonStream).catch(() => undefined),
+            extractImageLayer(layerStream, extractActions).catch(
+              () => undefined,
+            ),
+          ];
+          const [manifest, layer] = await Promise.all(promises);
+
+          // header format is /blobs/hash_name/hash_value
+          // we're extracting hash_name:hash_value format to match manifest digest
+          const headerParts = header.name.split("/");
+          const hashName = headerParts[1];
+          const hashValue = headerParts[headerParts.length - 1];
+          const digest = `${hashName}:${hashValue}`;
+          if (manifest !== undefined) {
+            manifests[digest] = manifest as OciArchiveManifest;
```

Comment:
I mean, something like this would be better, so I don't put files that don't care about into that object at all: 
```
          if (manifest !== undefined && manifest.layers) {
            manifests[digest] = manifest as OciArchiveManifest;
          }
```
But it errors at `manifest.layers` because the type is unknown at this point, I'm having a look

Reply:
Yup I think your idea is right! You can use the TS compiler to help you as well and create a type guard like this:
```js
function isOciArchiveManifest(manifest: any): manifest is OciArchiveManifest {
  return manifest !== undefined && manifest.layers && manifest.layers.length >= 0;
}
```
and then use it something like:
```js
if (isOciArchiveManifest(manifest)) {
  manifests[digest] = manifest; // as OciArchiveManifest; you don't need the cast, TS now knows it's of type OciArchiveManifest!
}
```